nkms/keystore/keystore.py

```diff
@@ -80,89 +58,90 @@ def get_key(self, fingerprint: bytes) -> Union[keypairs.EncryptingKeypair,
 
-    def add_key(self,
-                keypair: Union[keypairs.EncryptingKeypair,
-                               keypairs.SigningKeypair],
-                store_pub: bool = True) -> bytes:
+        :param fingerprint: Fingerprint of key to delete
         """
-        Gets a fingerprint of the key and adds it to the keystore.
+        self.session.query(Key).filter_by(fingerprint=fingerprint).delete()
+        self.session.commit()
 
-        :param key: Key, in bytes, to add to lmdb
+    def add_policy_contract(self, expiration, deposit, hrac, kfrag=None,
+                            alice_pubkey_sig=None, # alice_pubkey_enc,
+                            alice_signature=None) -> PolicyContract:
+        """
+        Creates a PolicyContract to the Keystore.
 
-        :return: Fingerprint, in bytes, of the added key
+        :return: The newly added PolicyContract object
         """
-        if store_pub:
-            fingerprint = self._get_fingerprint(keypair.pubkey)
-            key = keypair.serialize_pubkey()
-        else:
-            fingerprint = self._get_fingerprint(keypair.privkey)
-            key = keypair.serialize_privkey()
+        # TODO: This can be optimized to one commit/write.
+        alice_pubkey_sig = self.add_key(alice_pubkey_sig, is_signing=True)
+        # alice_pubkey_enc = self.add_key(alice_pubkey_enc)
+        # bob_pubkey_sig = self.add_key(bob_pubkey_sig)
 
-        # Create new Key object and commit to db
-        self.session.add(Key(key))
+        new_policy_contract = PolicyContract(
+            expiration, deposit, hrac, kfrag, alice_pubkey_sig=alice_pubkey_sig,
+            alice_signature=None, # bob_pubkey_sig.id
+        )
+
+        self.session.add(new_policy_contract)
         self.session.commit()
-        return fingerprint
 
-    def add_kfrag(self, hrac: bytes, kfrag: RekeyFrag, sig: bytes=None):
+        return new_policy_contract
+
+    def get_policy_contract(self, hrac: bytes) -> PolicyContract:
         """
-        Adds a RekeyFrag to sqlite.
+        Returns the PolicyContract by its HRAC.
 
-        :param hrac: Hashed Resource Authenticate Code
-        :param kfrag: RekeyFrag instance to add to sqlite
-        :param sig: Signature of kfrag (if exists)
+        :return: The PolicyContract object
         """
-        kfrag_data = b''
-        if sig:
-            kfrag_data += sig
-        kfrag_data += bytes(kfrag)
+        policy_contract = self.session.query(PolicyContract).filter_by(hrac=hrac).first()
+        if not policy_contract:
+            raise NotFound("No PolicyContract with {} HRAC found.".format(hrac))
+        return policy_contract
 
-        kfrag = KeyFrag(hrac, kfrag_data)
-        self.session.add(kfrag)
+    def del_policy_contract(self, hrac: bytes):
+        """
+        Deletes a PolicyContract from the Keystore.
+        """
+        self.session.query(PolicyContract).filter_by(hrac=hrac).delete()
         self.session.commit()
```

Comment:
Perhaps for a future PR - but we probably want to return (at minimum) the number of deleted records.

Reply:
👍

Reply:
✔️

Reply:
This will only delete a single `PolicyContract`. See the call to `.first()`.

Reply:
@tuxxy - Hmmm. not seeing a call to `.first()` in `del_policy_contract`. Looks like it bulk deletes all matching records from the `filter_by`.

Reply:
So where are we on this?  `hrac` is already unique on `PolicyContract`, so we know that we're deleting at most one here.  However, I also don't see a call to `.first()`.  

What do we need to do to make it ready for merge?